Web Inspector: Elements Tab: Computed: variables section looks broken when the selected element has no CSS variables

rcaliman-apple opened the pull request:
#### e069140a859cad8a65a868427fff4fdd6f1bf6aa
<pre>
Web Inspector: Elements Tab: Computed: variables section looks broken when the selected element has no CSS variables
<a href="https://bugs.webkit.org/show_bug.cgi?id=242704">https://bugs.webkit.org/show_bug.cgi?id=242704</a>

Reviewed by Devin Rousso.

The original implementation made some wrong assumptions about the layout lifecycle
for CSS variable groups. This patch introduces a refactoring to correct that and
fix a bunch issues at once.

Context:

The properties section of the Computed panel is updated in response to the event
`WI.CSSStyleDeclaration.Event.PropertiesChanged` fired by
`WI.DOMNodeStyles.computedStyle` which is a `WI.CSSDeclaration` instance.
This allows it to dynamically reflect changes to any styles of the selected node.

There&apos;s an optimization in `WI.DOMNodeStyles.refresh() -&gt; fetchedComputedStyle()`
where a change to styles of existing CSS rules isn&apos;t considered a `significantChange`.
This prevents a lot of needless re-layout in components which observe style refreshes.

The Properties section of the Computed panel listens for the event on `WI.DOMNodeStyles.computedStyle`
to address this optimization.

When grouping CSS variables, the groups were distinct `WI.CSSDeclaration` instances for clarity.
But they were cached and never updated in response to style changes on a node.
Because of this, the sections for variable groups in the Computed panel did not reflect
any changes: adding/removing variables, changing their type, etc.

The changes:

1) The logic for grouping variables by type is moved from `WI.DOMNodeStyles` to
`WI.CSSDeclaration`. This way, all sections in the Computed panel rely on a single
data source: `WI.DOMNodeStyles.computedStyle`.
It&apos;s the view&apos;s job to pick the subset of properties to render.

2) A new getter is added, `WI.CSSDeclaration.variablesForType`, to do this extra work
for all variable groups when consumers ask for them.

There can be a growing number of groups, so having separate iterations on every style
change over a potentially long list of properties is wasteful.
Plus, anything that doesn&apos;t match a type checker needs to go in the &quot;Other&quot; group.

3) `WI.ComputedStyleSection` is modified to obtain a list of properties to render
depending on the variables group type set on `WI.ComputedStyleSection.variablesGroupType`.
This subset of properties is then used in `WI.ComputedStyleSection.layout()`.

Introduced debouncing for layout in respose to `WI.DOMNodeStyles.Events.PropertiesChanged`
because it is expensive; throwing away all elements for properties and rebuilding from scratch.

4) Filtering is implemented. Sections that have no matches are hidden.

5) The test that checks CSS variables grouping is renamed and updated.

6) Modify `WI.View` to call `WI.View.didSubtreeLayout()` on the immediate parent
of the first dirty view that did layout on itself and all of its descendant views.
This allows a parent view to react in response to layout under its subtree.

This behavior is used to call `WI.ComputedStyleDetailsPanel.didSubtreeLayout()` where we can
toggle the visibility of the &quot;Variables&quot; section when the nested sections for variable groups are empty.

* LayoutTests/inspector/css/variablesForType-expected.txt: Renamed from LayoutTests/inspector/css/variableStylesByType-expected.txt.
* LayoutTests/inspector/css/variablesForType.html: Renamed from LayoutTests/inspector/css/variableStylesByType.html.
* LayoutTests/inspector/view/asynchronous-layout-expected.txt:
* LayoutTests/inspector/view/asynchronous-layout.html:
* LayoutTests/inspector/view/resources/test-view.js:
(TestPage.registerInitializer.WI.TestView):
(TestPage.registerInitializer.WI.TestView.prototype.get didLayoutSubtreeCount):
(TestPage.registerInitializer.WI.TestView.prototype.didLayoutSubtree):
(TestPage.registerInitializer):
* LayoutTests/inspector/view/synchronous-layout-expected.txt:
* LayoutTests/inspector/view/synchronous-layout.html:
* Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:
(WI.CSSStyleDeclaration):
(WI.CSSStyleDeclaration.prototype.variablesForType):
* Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
(WI.DOMNodeStyles):
(WI.DOMNodeStyles.prototype.refresh.fetchedComputedStyle):
(WI.DOMNodeStyles.prototype.get variableStylesByType): Deleted.
* Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
(WI.ComputedStyleDetailsPanel.displayNameForVariablesGroupType):
(WI.ComputedStyleDetailsPanel.prototype.refresh):
(WI.ComputedStyleDetailsPanel.prototype.layout):
(WI.ComputedStyleDetailsPanel.prototype.didLayoutSubtree):
(WI.ComputedStyleDetailsPanel.prototype._createVariablesStyleSection):
(WI.ComputedStyleDetailsPanel.prototype._removeVariablesStyleSection):
(WI.ComputedStyleDetailsPanel.prototype._renderVariablesStyleSectionGroup): Deleted.
* Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js:
(WI.ComputedStyleSection.prototype.get variablesGroupType):
(WI.ComputedStyleSection.prototype.get isEmpty):
(WI.ComputedStyleSection.prototype.get properties):
(WI.ComputedStyleSection.prototype.layout):
(WI.ComputedStyleSection.prototype.detached):
(WI.ComputedStyleSection.prototype.applyFilter):
(WI.ComputedStyleSection.prototype._handlePropertiesChanged):
(WI.ComputedStyleSection):
* Source/WebInspectorUI/UserInterface/Views/View.js:
(WI.View.prototype.updateLayout):
(WI.View.prototype._layoutSubtree):
(WI.View._visitViewTreeForLayout):
(WI.View):

Canonical link: <a href="https://commits.webkit.org/253562@main">https://commits.webkit.org/253562@main</a>
</pre>















<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3ff86a6dc24a74b69f5a519222e8b8c994307d32

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86401 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30321 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17369 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95249 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/148957 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/28686 "Built successfully") | [✅ ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/78552 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90493 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92002 "Passed tests") | [✅ ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/23266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/73381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23358 "Passed tests") | 
| | [✅ ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/78278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/78705 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/66364 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/26636 "Built successfully") | [✅ ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/12550 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/26549 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/13565 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2535 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28227 "Built successfully") | [✅ ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/73381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28167 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/32847 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js:
```diff
@@ -221,6 +244,14 @@ WI.ComputedStyleSection = class ComputedStyleSection extends WI.View
+        if (this._filterCleared)
+            this._filterCleared = false;
+
+        this.dispatchEventToListeners(WI.ComputedStyleSection.Event.LayoutUpdated, {isEmpty: this.isEmpty})
```
This seems kinda like a layering violation.  Can we use `didLayoutSubtree` instead?

rcaliman-apple replied:
Good idea! Replaced dispatching an event with `didLayoutSubtree()`. Neat! I forgot about this.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js:
```diff
@@ -221,6 +244,14 @@ WI.ComputedStyleSection = class ComputedStyleSection extends WI.View
+            this.dispatchEventToListeners(WI.ComputedStyleSection.Event.FilterApplied, {matches: !this.isEmpty});
+
+        if (this._filterCleared)
+            this._filterCleared = false;
```
why not just this?
```
this._filterCleared = false;
```

rcaliman-apple replied:
For clarity, really. But the standalone assignment works the same. Fixed.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js:
```diff
@@ -129,16 +130,38 @@ WI.ComputedStyleSection = class ComputedStyleSection extends WI.View
+            return this._propertiesGetter(this._style);
+
+        if (this._style._styleSheetTextRange)
+           return this._style.visibleProperties;
```
Style: odd number of spaces

rcaliman-apple replied:
Fixed

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js:
```diff
@@ -129,16 +130,38 @@ WI.ComputedStyleSection = class ComputedStyleSection extends WI.View
 
+    get isEmpty()
+    {
+        this.updateLayoutIfNeeded();
```
oof is there no way to avoid this?  we shouldn't do sync forced layout unless absolutely necessary

rcaliman-apple replied:
Removed the forced layout. 
This was a chunk in plucked from another failed patch where `isEmpty()` could have been called before the section did layout. Not the case here.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
```diff
@@ -303,6 +286,23 @@ WI.ComputedStyleDetailsPanel = class ComputedStyleDetailsPanel extends WI.StyleD
+        detailsSection?.element.classList.toggle("hidden", event.data.isEmpty);
+
+        let areVariableSectionsAllEmpty = true;
+        for (let [styleSection, detailsSection] of this._detailsSectionByStyleSectionMap) {
```
```
for (let styleSection of this._detailsSectionByStyleSectionMap.keys())
```

rcaliman-apple replied:
Done.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
```diff
@@ -267,6 +246,9 @@ WI.ComputedStyleDetailsPanel = class ComputedStyleDetailsPanel extends WI.StyleD
         variablesStyleSection.style = style;
 
+        if (groupType !== WI.CSSStyleDeclaration.VariablesGroupType.Ungrouped)
+            variablesStyleSection.propertiesGetter = (style) => { return style.getVariablesForType(groupType) }
```
```
variablesStyleSection.propertiesGetter = (style) => style.getVariablesForType(groupType);
```

rcaliman-apple replied:
Dropped the custom getter in favor of your other suggestion to make `WI.ComputedStyleSection` grab its own subset of properties based on the variable group type.

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:
```diff
@@ -113,6 +115,22 @@ WI.CSSStyleDeclaration = class CSSStyleDeclaration extends WI.Object
     get locked() { return this._locked; }
     set locked(value) { this._locked = value; }
 
+    get useVariablesGrouping() { return this._useVariablesGrouping; }
```
Style: we only do single line `get` or `set` if both the `get` and `set` are single line (or if there is only a `get`)

rcaliman-apple replied:
Fixed.

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:
```diff
@@ -113,6 +115,22 @@ WI.CSSStyleDeclaration = class CSSStyleDeclaration extends WI.Object
+            this.update(this.text, this.properties, this.styleSheetTextRange, {dontFireEvents: true});
+    }
+
+    getVariablesForType(type)
```
I think we could/should just name this `variablesForType`

rcaliman-apple replied:
Done.

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:
```diff
@@ -113,6 +115,22 @@ WI.CSSStyleDeclaration = class CSSStyleDeclaration extends WI.Object
+
+        // Force an update with the same parameters to trigger variable grouping.
+        if (wasUngrouped)
+            this.update(this.text, this.properties, this.styleSheetTextRange, {dontFireEvents: true});
```
Why is this necessary?  Could we instead generate this lazily inside `variablesForType` and then clear the `Map` inside `update`?

_reads commit message_

I dont really think the performance of iterating `_properties` is likely to be of concern as I'd be very surprised if there were more than a thousand (and probably not even anything close to that) CSS variables at the same time.  Not to mention, we're effectively forcing a second iteration when switching to the Computed panel anyways, so I'm not sure what much we gain by doing this.

rcaliman-apple replied:
> Could we instead generate this lazily inside variablesForType 

Yes. That's a good idea. Updated.

> I dont really think the performance of iterating _properties is likely to be of concern

`DOMNodeStyles.computedStyle.update()` is [called every time](https://sourcegraph.com/github.com/WebKit/WebKit-http@817c46e152af795d735678386db68805d0aa505e/-/blob/Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js?L248) there's a change to the styles of the selected node. Typing a CSS property name or value means a flurry of activity. It's enough to do just one iteration over the list of properties in the computed style. 

I'd argue we could throttle or debounce the handler for the `WI.CSSStyleDeclaration.Event.PropertiesChanged` event. The Computed style panel doesn't need to be _live_ in sync with the Styles panel.

What concerns me more is that every single style change thrashes the DOM in`WI.ComputedStyleSection.layout()` by [removing all elements](https://sourcegraph.com/github.com/WebKit/WebKit-http@817c46e152af795d735678386db68805d0aa505e/-/blob/Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js?L197) and rebuilding from scratch. This is very wasteful. I'd like to introduce a `WeakMap` that matches incoming properties with previously created `WI.SpreadsheetStyleProperty` instances, use it to check what changed, and update the DOM accordingly. But that's going to happen in a follow-up. This patch is already sprawling.

dcrousso replied:
apologies, i dont think i was entirely clear

what i was really trying to get at is why are we doing this inside `update` at all?  why not just iterate over `_properties` inside `variablesForType`?
```
variablesForType(type)
{
    if (!this._variablesForType) {
        this._variablesForType = new Map;
        for (let property of this._properties) {
            if (property.isVariable)
                continue;

            ...
        } 
    }
    return this._variablesForType;
}
```
(and we'd clear `this._variablesForType` inside `update`)

this would also let you drop the `useVariablesGrouping` concept as anyone could call it (and really thats probably how it should be too)

calling `update` feels a bit heavy cause it also adjusts things like `this._propertyNameMap` and the `WI.CSSProperty.prototype.get index`, which IMO is entirely unnecessary

as far as the other things, i definitely agree that we should improve (in a followup) how we handle changes to the computed style :)  throttling would be a nice start for sure

rcaliman-apple replied:
Thought about it some more and I agree. The separation wasn't necessary.
Changed the patch to do variable grouping lazily in `this.variablesForType()`.
Dropped `useVariablesGrouping`.

Now, we'll generate a `Map` with lists of variables for every group type. Empty lists for groups with no matches serve as a signal for `WI.ComputedStyleDetailsPanel` to avoid rendering sections that won't have content. This shaves some more unnecessary loops for sections known to be empty.

rcaliman-apple replied:
Modified `WI.View` to call `WI.View.didLayoutSubtree()` on the immediate parent
of the first dirty view that did layout on itself and all of its descendant views.

rcaliman-apple replied:
Used @dcrousso's patch to modify `WI.View._visitViewTreeForLayout()`.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js:
```diff
@@ -129,16 +130,38 @@ WI.ComputedStyleSection = class ComputedStyleSection extends WI.View
+    {
+        console.assert(typeof fn === "function", fn, "The custom getter must be a function that returns an array of properties.");
+
+        this._propertiesGetter = fn;
```
something about this feels icky

we already are creating a separate `WI.ComputedStyleSection` for each `WI.CSSStyleDeclaration.VariablesGroupType` (via `WI.ComputedStyleDetailsPanel.prototype._renderVariablesStyleSectionGroup`) , so could we instead pass the `WI.CSSStyleDeclaration.VariablesGroupType` to the `WI.ComputedStyleSection` and then have `WI.ComputedStyleSection.prototype.get properties` use it (if provided)?

rcaliman-apple replied:
I like this idea. Implemented. Thanks!

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js:
```diff
@@ -221,6 +248,16 @@ WI.ComputedStyleSection = class ComputedStyleSection extends WI.View
+
+    didLayoutSubtree()
+    {
+        this._delegate.computedStyleSectionLayoutUpdated?.(this, this.isEmpty);
```
Are we guaranteed to have a `this._delegate`?  Other calls don't seem to think so :/

Also, `didLayoutSubtree` should get called for the parent too.  Do we even need a delegate call?

rcaliman-apple replied:
> Are we guaranteed to have a `this._delegate`

Yes, `WI.ComputedStyleSection` always has a delegate. Instances are created only in `WI.ComputedStyleDetailsPanel` which is the delegate passed down.

> Also, didLayoutSubtree should get called for the parent too. Do we even need a delegate call?

We still need the delegate call. I kind of prefer it.

This part suffers from the weirdness that a `WI.View` (`WI.ComputedStyleSection`) has its element manually  attached to a `WI.Object` (`WI.DetailsSectionRow`). It's not part of a continuous view tree between `WI.ComputedStyleSection` and `WI.ComputedStyleDetailsPanel`. I briefly poked it to try and make use of `didLayoutSubtree()` on the ancestor view, but it didn't work right away and I didn't investigate further. 

Given the arcane setup, I think calling `this._delegate.computedStyleSectionLayoutUpdated()` is more clear.

dcrousso replied:
> > Are we guaranteed to have a `this._delegate`
> 
> Yes, `WI.ComputedStyleSection` always has a delegate. Instances are created only in `WI.ComputedStyleDetailsPanel` which is the delegate passed down.

in that case we probably should add a `console.assert(delegate)` in the constructor

> > Also, didLayoutSubtree should get called for the parent too. Do we even need a delegate call?
> 
> We still need the delegate call. I kind of prefer it.
> 
> This part suffers from the weirdness that a `WI.View` (`WI.ComputedStyleSection`) has its element manually attached to a `WI.Object` (`WI.DetailsSectionRow`). It's not part of a continuous view tree between `WI.ComputedStyleSection` and `WI.ComputedStyleDetailsPanel`. I briefly poked it to try and make use of `didLayoutSubtree()` on the ancestor view, but it didn't work right away and I didn't investigate further.
> 
> Given the arcane setup, I think calling `this._delegate.computedStyleSectionLayoutUpdated()` is more clear.

i'm not sure i understand

we call `this.addSubview` with the `WI.ComputedStyleSection`, meaning that it should participate in the layout system.  the fact that it's a child of a `WI.DetailsSectionRow` in the DOM only affects where it appears in the UI, not whether it participates in the layout system (which it still does).

if you're not seeing `didLayoutSubtree` get called for the `WI.ComputedStyleDetailsPanel` after any `WI.ComputedStyleSection` does a `layout` then we have a bug in `WI.View` that absolutely should be fixed, because `didLayoutSubtree` should _always_ be called after the subtree finishes layout

rcaliman-apple replied:
> if you're not seeing didLayoutSubtree get called for the WI.ComputedStyleDetailsPanel after any WI.ComputedStyleSection does a layout then we have a bug in WI.View that absolutely should be fixed, because didLayoutSubtree should always be called after the subtree finishes layout

I don't think this is a bug.

Looking at `WI.View._visitViewTreeForLayout()`, layout is triggered only on subviews which are "dirty" (layout pending):

[Source/WebInspectorUI/UserInterface/Views/View.js?L376-378](https://sourcegraph.com/github.com/WebKit/WebKit@29b13de20953a2bdfdb1f5dfa7bf3c856d120009/-/blob/Source/WebInspectorUI/UserInterface/Views/View.js?L376-378)
```
for (let i = 0; i < views.length; ++i) {
    let view = views[i];
    if (view.layoutPending)
        view._layoutSubtree();
    else if (view._dirtyDescendantsCount)
        views.pushAll(view.subviews);
}
```

Once layout is done, the top-most subview that was "dirty" and all its subviews will indeed call `this.didLayoutSubtree()`.

[Source/WebInspectorUI/UserInterface/Views/View.js?L317-326](https://sourcegraph.com/github.com/WebKit/WebKit@29b13de20953a2bdfdb1f5dfa7bf3c856d120009/-/blob/Source/WebInspectorUI/UserInterface/Views/View.js?L317-326)

```
for (let view of this._subviews) {
    view._setLayoutReason(this._layoutReason);
    view._layoutSubtree();
}

this._layoutReason = null;

console.assert(WI.setReentrantCheck(this, "didLayoutSubtree"), "ERROR: calling `didLayoutSubtree` while already in it", this);
this.didLayoutSubtree();
```

In our case, `WI.ComputedStyleSection` is the top-most subview that does layout in response to handling the event `WI.CSSStyleDeclaration.Event.PropertiesChanged` from the computed styles object. 

The parent `WI.ComputedStyleDetailsPanel` doesn't do layout in response to this event. It does layout only when switching nodes or a `significantChange` occurs in the selected node's style.

That's why `WI.ComputedStyleDetailsPanel.didSubtreeLayout()` doesn't get called. This view isn't "dirty", but its children are.

dcrousso replied:
just because the code is currently written that way does not mean that it is incorrect

the whole idea of `didLayoutSubtree` is so that parents can do extra (ideally non-layout) work after all children have done their `layout`, meaning that the entire subtree is in a known "good" state (see the existing usage of `didLayoutSubtree` for other examples of this desired behavior)

as such, we should be calling `didLayoutSubtree` on the parent even if it's not `dirty` (i.e. it did not call `layout` on itself) so that parents can respond to changes in their children

heck one of the reasons `didLayoutSubtree` was added was to avoid this exact scenario of using an event/delegate/etc. to signal "hey im done with `layout`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
```diff
@@ -267,6 +264,9 @@ WI.ComputedStyleDetailsPanel = class ComputedStyleDetailsPanel extends WI.StyleD
         variablesStyleSection.style = style;
 
+        if (groupType !== WI.CSSStyleDeclaration.VariablesGroupType.Ungrouped)
+            variablesStyleSection.variablesGroupType = groupType;
```
Why not just provide this on construction of the `WI.ComputedStyleSection`?  I feel like it should be the one to decide what to do with `WI.CSSStyleDeclaration.VariablesGroupType.Ungrouped`, not some UI class/logic.  This would also let you drop the `get variablesGroupType` and `set variablesGroupType`.

rcaliman-apple replied:
Yeah, that's a good idea. Moved `variablesGroupType` as an option passed to the constructor. 

I intentionally didn't move other properties for `WI.ComputedStyleSection` to the new options object.
Perhaps I'll follow up on that in another patch. Trying to keep this one focused.

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:
```diff
@@ -145,6 +193,7 @@ WI.CSSStyleDeclaration = class CSSStyleDeclaration extends WI.Object
 
         this._styleSheetTextRange = styleSheetTextRange;
         this._propertyNameMap = {};
+        this._variablesForType = new Map;
```
NIT: if we always expect it to be a `Map` (i.e. it's never `null`) then i think we could/should just `.clear()` it

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:
```diff
@@ -113,6 +114,53 @@ WI.CSSStyleDeclaration = class CSSStyleDeclaration extends WI.Object
+    {
+        console.assert(Object.values(WI.CSSStyleDeclaration.VariablesGroupType).includes(type), type, "Unknown variables group type.");
+
+        if (this._variablesForType.has(type))
```
NIT: would be nice to adjust this code so that we don't have to do so many lookups

one example is that we probably don't have to check `.has(type)` because we never insert anything other than an array, so checking if the value is falsy should be enough
```
let variables = this._variablesForType.get(type);
if (variables)
    return variables;

...
```

rcaliman-apple replied:
Isn't a lookup by key in a map O(1)?

dcrousso replied:
that's not what ive been told or what it looks like based on `HashMapImpl` (which `JSMap` inherits from)

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:
```diff
@@ -113,6 +114,53 @@ WI.CSSStyleDeclaration = class CSSStyleDeclaration extends WI.Object
     get locked() { return this._locked; }
     set locked(value) { this._locked = value; }
 
+    variablesForType(type)
```
can we add a test for this?  i think the old test would work, just adjusting it to use `WI.CSSStyleDeclaration` instead of `WI.DOMNodeStyles`

rcaliman-apple replied:
Yes. I had modified and renamed the old test. It somehow got dropped in my `git` juggling. Will add it back.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
```diff
@@ -89,47 +89,41 @@ WI.ComputedStyleDetailsPanel = class ComputedStyleDetailsPanel extends WI.StyleD
+        this._propertiesSection.element.classList.toggle("hidden", !this._computedStyleSection.propertiesToRender.length);
+        this._boxModelDiagramRow.nodeStyles = this.nodeStyles;
 
-        super.refresh();
```
why not still call `super.refresh(significantChange)`?

rcaliman-apple replied:
Added back, though the super class doesn't do anything with it.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
```diff
@@ -89,47 +89,41 @@ WI.ComputedStyleDetailsPanel = class ComputedStyleDetailsPanel extends WI.StyleD
+                    let variablesList = this.nodeStyles.computedStyle.variablesForType(type);
+                    if (variablesList.length && !this._variablesStyleSectionForGroupTypeMap.has(type))
+                        this._createVariablesStyleSection(type, WI.ComputedStyleDetailsPanel.displayNameForVariablesGroupType(type));
+                    else if (!variablesList.length && this._variablesStyleSectionForGroupTypeMap.has(type))
```
ditto (Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:+121)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
```diff
@@ -151,6 +145,17 @@ WI.ComputedStyleDetailsPanel = class ComputedStyleDetailsPanel extends WI.StyleD
+        let detailsSection = this._detailsSectionByStyleSectionMap.get(styleSection);
+        detailsSection?.element.classList.toggle("hidden", isEmpty);
+
+        let areVariableSectionsAllEmpty = Array.from(this._variablesStyleSectionForGroupTypeMap.values()).every(styleSection => styleSection.isEmpty);
```
Style: we always add `(` `)` even for single argument arrow functions

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
```diff
@@ -151,6 +145,17 @@ WI.ComputedStyleDetailsPanel = class ComputedStyleDetailsPanel extends WI.StyleD
     }
 
+    // ComputedStyleSection delegate
+    computedStyleSectionLayoutUpdated(styleSection, isEmpty)
```
Style: missing newline in between

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
```diff
@@ -151,6 +145,17 @@ WI.ComputedStyleDetailsPanel = class ComputedStyleDetailsPanel extends WI.StyleD
+        this._variablesSection.element.classList.toggle("hidden", areVariableSectionsAllEmpty);
+    }
+
+
```
Style: extra newline

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
```diff
@@ -225,25 +230,24 @@ WI.ComputedStyleDetailsPanel = class ComputedStyleDetailsPanel extends WI.StyleD
         }
 
-        this._variablesStyleSectionForGroupTypeMap.clear();
+        if (!this.nodeStyles.computedStyle)
```
Is this actually possible?  Or is this a proxy for checking `this.nodeStyles.needsRefresh`?

rcaliman-apple replied:
`WI.DOMNodeStyles.computedStyle` starts off as `null`. It gets populated in a promise:
[Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js?L320-323](https://sourcegraph.com/github.com/WebKit/WebKit@29b13de20953a2bdfdb1f5dfa7bf3c856d120009/-/blob/Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js?L320-323)

I guess checking `this.nodeStyles.needsRefresh` would also be fine.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
```diff
@@ -258,26 +262,47 @@ WI.ComputedStyleDetailsPanel = class ComputedStyleDetailsPanel extends WI.StyleD
+    _createVariablesStyleSection(variablesGroupType, label)
     {
-        let variablesStyleSection = new WI.ComputedStyleSection(this);
+        let variablesStyleSection = new WI.ComputedStyleSection(this, { variablesGroupType });
```
Style: no space after `{` or before `}`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
```diff
@@ -258,26 +262,47 @@ WI.ComputedStyleDetailsPanel = class ComputedStyleDetailsPanel extends WI.StyleD
+
+    _removeVariablesStyleSection(styleSection)
+    {
+        console.assert(this._detailsSectionByStyleSectionMap.has(styleSection), styleSection, "Expected a ComputedStyleSection for a variable group")
```
ditto (Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:+121)

Style: missing `;`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
```diff
@@ -258,26 +262,47 @@ WI.ComputedStyleDetailsPanel = class ComputedStyleDetailsPanel extends WI.StyleD
+        this.element.appendChild(styleSection.element);
+        this.removeSubview(styleSection);
+
+        styleSection.element.remove();
```
this should be handled by `removeSubview`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
```diff
@@ -258,26 +262,47 @@ WI.ComputedStyleDetailsPanel = class ComputedStyleDetailsPanel extends WI.StyleD
+
+        // WI.ComputedStyleSection is a view but its element is attached to a WI.DetailsSection which isn't. Reparent it before removing.
+        // FIXME: <https://webkit.org/b/152269> - Web Inspector: Convert DetailsSection classes to use View
+        this.element.appendChild(styleSection.element);
```
at this point is the `detailsSection` attached to `this.element`?  if so, then this is unnecessary as `WI.View.prototype.removeSubview` only checks that the `WI.View` is in the subtree, not that it's a direct descendant

rcaliman-apple replied:
That's true. The DOM juggling isn't necessary anymore.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
```diff
@@ -258,26 +262,47 @@ WI.ComputedStyleDetailsPanel = class ComputedStyleDetailsPanel extends WI.StyleD
+        detailsSection.element.remove();
+        detailsSection.removeEventListener(WI.DetailsSection.Event.CollapsedStateChanged, this._handleDetailsSectionCollapsedStateChanged, this);
+
+        this._detailsSectionByStyleSectionMap.delete(styleSection);
```
let's use `.take` above instead

rcaliman-apple replied:
Why? We don't need the value.

dcrousso replied:
you could use it inside the `console.assert` instead of the `.has`.  that's typically how we do things like this

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
```diff
@@ -258,26 +262,47 @@ WI.ComputedStyleDetailsPanel = class ComputedStyleDetailsPanel extends WI.StyleD
+        detailsSection.removeEventListener(WI.DetailsSection.Event.CollapsedStateChanged, this._handleDetailsSectionCollapsedStateChanged, this);
+
+        this._detailsSectionByStyleSectionMap.delete(styleSection);
+        this._variablesStyleSectionForGroupTypeMap.delete(styleSection.variablesGroupType);
```
should we also `console.assert` that this is true?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js:
```diff
@@ -25,8 +25,10 @@
-    constructor(delegate)
+    constructor(delegate, {variablesGroupType} = {})
     {
+        console.assert(!variablesGroupType || Object.values(WI.CSSStyleDeclaration.VariablesGroupType).includes(variablesGroupType), "Unknown variables group type.");
```
please just log the `variablesGroupType` instead of having a message

```
console.assert(!variablesGroupType || Object.values(WI.CSSStyleDeclaration.VariablesGroupType).includes(variablesGroupType), variablesGroupType);
```
anyone reading this code will immediately be able to understand what the problem is, and knowing _what_ the invalid value is is often more useful than generally knowing that there is an invalid value

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js:
```diff
@@ -129,16 +133,28 @@ WI.ComputedStyleSection = class ComputedStyleSection extends WI.View
         this.needsLayout();
     }
 
+    get variablesGroupType() { return this._variablesGroupType; }
```
please move one-line getters to the beginning of the `// Public` section

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js:
```diff
@@ -129,16 +133,28 @@ WI.ComputedStyleSection = class ComputedStyleSection extends WI.View
 
+    get variablesGroupType() { return this._variablesGroupType; }
+
+    get isEmpty() {return !this._propertyViews.length; }
```
Style: this is not really a simple getter (i.e. it returns a member variable of the same name), so please expand this onto multiple lines

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js:
```diff
@@ -129,16 +133,28 @@ WI.ComputedStyleSection = class ComputedStyleSection extends WI.View
+
+        if (this._style._styleSheetTextRange)
+            return this._style.visibleProperties;
+        else
```
Style: the `else` is not necessary because the `if` contains a `return`
```
if (this._style.styleSheetTextRange)
    return this._style.visibleProperties;
return this._style.properties;
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js:
```diff
@@ -129,16 +133,28 @@ WI.ComputedStyleSection = class ComputedStyleSection extends WI.View
+        if (this._variablesGroupType && this._variablesGroupType !== WI.CSSStyleDeclaration.VariablesGroupType.Ungrouped)
+            return this._style.variablesForType(this._variablesGroupType);
+
+        if (this._style._styleSheetTextRange)
```
can we use `this._style.styleSheetTextRange` instead?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js:
```diff
@@ -297,7 +330,10 @@ WI.ComputedStyleSection = class ComputedStyleSection extends WI.View
     {
-        this.needsLayout();
+        if (!this._propertiesChangedDebouncer)
+            this._propertiesChangedDebouncer = new Debouncer(this.needsLayout.bind(this));
```
```
this._propertiesChangedDebouncer ||= new Debouncer(this.needsLayout.bind(this));
this._propertiesChangedDebouncer.delayForTime(250);
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/View.js:
```diff
@@ -285,6 +285,7 @@ WI.View = class View extends WI.Object
     {
         this._setDirty(false);
         let isInitialLayout = !this._didInitialLayout;
+        let hasNonDirtyParentView = !this.parentView.layoutReason && !this.parentView.dirty;
```
I think we also need to check the `_dirtyDescendantsCount` as that's really what determines whether this is the last child `WI.View` in need of a `layout`.  the problem with that, however, is that `_dirtyDescendantsCount` keeps track of the entire subtree, not just the direct children, so you'd have to check that after calling `layout`, but there's also nothing stopping a `WI.View` from calling `needsLayout` inside `layout`, so it's possible that even looking at `_dirtyDescendantsCount` won't be a good signal.

we might need to add a separate `_dirtyChildCount` that keeps track of just the direct children (and then maybe rename `_dirtyDescendantsCount` to be more specific to the fact that it's about grandchildren and below) and then check to see when it's equal to `1` (i.e. "this is the last dirty child") before we `this._setDirty(false)`

also please add some tests for this in `LayoutTests/inspector/view/` :)

rcaliman-apple replied:
> I think we also need to check the `_dirtyDescendantsCount` as that's really what determines whether this is the last child `WI.View` in need of a `layout`.

I've been thinking about this and looking at uses. 

We're not trying to determine the **last** `WI.View` in need of a `layout`. We're looking for the **parent of the first** `WI.View` in need of a `layout`.

I don't think we need an additional check other than a missing `WI.View.layoutReason`. 

```diff
- let hasNonDirtyParentView = !this.parentView.layoutReason && !this.parentView.dirty;
+ let hasNonDirtyParentView = !this.parentView?.layoutReason;
```

Looking at the dirty state of the parent view when doing layout for the subview is moot because it always `false`.

Here's how I understand layout of view tree happens:

`WI.View._layoutSubtree()` gets called from one of two paths:
- directly when calling `WI.View.updateLayout()`
- indirectly when scheduling with`WI.View.needsLayout()` -->  `WI.View._scheduleLayoutForView()`

In both cases, the layout happens from top to bottom along the view tree.
When called directly, it starts from a specific view.
When scheduled, it starts from `_rootView`. 

When called directly, a `WI.View._layoutReason` is set, either explicitly or fall back to `WI.View.LayoutReason.Dirty`.
When scheduled, a dirty flag is set: `WI.View._setDirty(true)`. **Here we can also set `view._setLayoutReason(WI.View.LayoutReason.Dirty)`**. 

When layout starts from the top-most view that needs layout, the first thing that happens in `WI.View._layoutSubtree()` is to flip its own dirty flag, `this._setDirty(false)`, then call `this.layout()`, then recursively call `view._layoutSubtree()` on all its subviews (while passing down the `_layoutReason`), and **finally reset the `_layoutReason` to null**.

So by the time a subview does layout, its parent's `_layoutReason` hasn't been reset yet. 
And _all_ subviews will have `_layoutSubtree()` called on them.

From this behavior we can infer that the parent view which doesn't have a truthy `_layoutReason` was not `dirty`. That's our target parent view for which we want to call `didLayoutSubtree()`.

The modifications that I need to make are:
```diff
_layoutSubtree()
{
-   let hasNonDirtyParentView = !this.parentView.layoutReason && !this.parentView.dirty;
+   let hasNonDirtyParentView = !this.parentView?.layoutReason;
...
}

static _scheduleLayoutForView(view)
{
...
   view._setDirty(true);
+  view._setLayoutReason(WI.View.LayoutReason.Dirty);
...
```

I believe this handles most cases, including: 
> there's also nothing stopping a WI.View from calling needsLayout inside layout,
because that goes via the scheduling route, where we'll set the layout reason.

Is there any use case I'm missing from this logic?
If it works, I prefer this over keeping counts of child views.

dcrousso replied:
> We're not trying to determine the **last** `WI.View` in need of a `layout`. We're looking for the **parent of the first** `WI.View` in need of a `layout`.

But what about if two children of the same `WI.View` are dirty?  We only want to call `didLayoutSubtree` after all children of a given `WI.View` are done with their `layout` (for a single `updateLayout`/`WI.View._visitViewTreeForLayout`).  Calling `didLayoutSubtree` after the first child `WI.View` means that the parent could be looking at incorrect state for any of its other child `WI.View` that would then update when they do their `layout`.

Consider this example

```
WI.CounterView = class CounterView extends WI.View
{
    layout()
    {
        this.count = (this.count || 0) + 1;
    }
};

WI.ParentView = class ParentView extends WI.View
{
    initialLayout()
    {
        this.addSubview(new WI.CounterView);
        this.addSubview(new WI.CounterView);
        this.addSubview(new WI.CounterView);
    }

    didLayoutSubtree()
    {
        console.log(this.subviews.map((subview) => subview.count).join(" "));
    }
};

let parentView = new WI.ParentView;
WI.View.rootView().addSubview(parentView);
WI.View._visitViewTreeForLayout();

parentView.subviews[1].needsLayout();
parentView.subviews[2].needsLayout();
WI.View._visitViewTreeForLayout();
```

If `didLayoutSubtree` is called after the first child view does its `layout`, then we'd see `1 2 1` get logged.  But the correct output should be `1 2 2` since both the 2nd and the 3rd child views are about to `layout`.

To put it another way and/or (re)summarize, we need a way of knowing "does my parent have any other children that need to `layout`?" in order to correctly call `didLayoutSubtree`.

> Looking at the dirty state of the parent view when doing layout for the subview is moot because it always `false`.

Oh good catch!  I forgot we set it to `false` super early.  We could probably adjust that if needed, but i think your idea of using the `layoutReason` is a bit easier.

> ```
> static _scheduleLayoutForView(view)
> {
> ...
>    view._setDirty(true);
> +  view._setLayoutReason(WI.View.LayoutReason.Dirty);
> ...
> ```

Hmm the only way I can see `layout` getting called without having a `layoutReason` is if the `WI.View` was just added as a subview (which is definitely something we should fix).  All other paths (e.g. `needsLayout`) should already be setting a `layoutReason`.  Is there some other path I'm unaware of?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
```diff
@@ -225,59 +221,85 @@ WI.ComputedStyleDetailsPanel = class ComputedStyleDetailsPanel extends WI.StyleD
+    {
+        super.didLayoutSubtree();
+
+        for (let [styleSection, detailsSection] of this._detailsSectionByStyleSectionMap) {
```
Style: no `{` `}` around single line control flow

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
```diff
@@ -225,59 +221,85 @@ WI.ComputedStyleDetailsPanel = class ComputedStyleDetailsPanel extends WI.StyleD
+        super.didLayoutSubtree();
+
+        for (let [styleSection, detailsSection] of this._detailsSectionByStyleSectionMap) {
+            detailsSection?.element.classList.toggle("hidden", styleSection.isEmpty);
```
is it actually possible for `detailsSection` to be `null`/`undefined`?  i'd think that everything in the `Map` would always exist

rcaliman-apple replied:
Yes, that's true. This leftover from copy/pasta

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js:
```diff
@@ -43,10 +46,14 @@ WI.ComputedStyleSection = class ComputedStyleSection extends WI.View
         this._hideFilterNonMatchingProperties = false;
         this._filterText = null;
+        this._filterCleared = false;
+        this._variablesGroupType = variablesGroupType;
```
Style: since this is optional, we should have a `|| null` or something to make sure we don't end up with an `undefined`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js:
```diff
@@ -221,18 +240,28 @@ WI.ComputedStyleSection = class ComputedStyleSection extends WI.View
         }
+
+        if (this._filterText || this._filterCleared)
+            this.dispatchEventToListeners(WI.ComputedStyleSection.Event.FilterApplied, {matches: !this.isEmpty});
```
is this actually necessary anymore now that we use `didLayoutSubtree`?

rcaliman-apple replied:
It's not needed anymore in this implementation, but I'd like to keep it for now. 

It's wasteful that filtering triggers a section layout which blows away the entire list of properties. In a future patch, I want to optimize this so filtering should be a cheaper operation. Something like `WI.SpreadsheetCSSStyleDeclarationEditor.applyFilter()`. Then, dispatching the event will occur independent of layout.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js:
```diff
@@ -129,16 +136,28 @@ WI.ComputedStyleSection = class ComputedStyleSection extends WI.View
         this.needsLayout();
     }
 
+    get isEmpty() {
```
Style: `{` should be on it's own line

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/View.js:
```diff
@@ -157,6 +157,7 @@ WI.View = class View extends WI.Object
     {
         this._setLayoutReason(layoutReason);
         this._layoutSubtree();
+        this.parentView?.didLayoutSubtree();
```
NIT: Id use `this._parentView` since we're inside the same class and can therefore avoid having to invoke a `get`

rcaliman-apple replied:
Fixed

dcrousso commented on LayoutTests/inspector/view/asynchronous-layout-expected.txt:
```diff
@@ -31,22 +31,31 @@ PASS: Parent view should have 0 dirty descendants.
-PASS: Parent view should have 0 dirty descendants.
-PASS: Child view should have 0 dirty descendants.
-PASS: Parent view should have started a layout.
-PASS: Child view should have completed 2 layouts.
```
I think this is because of your changes to `WI.View.prototype.updateLayout`, but looking at the test I'm slightly concerned because the test is supposed to end after "Parent view completed a layout.", so we should either update the test or maybe even rework how `WI.TestView.prototype.evaluateAfterLayout` works 🤔

rcaliman-apple replied:
Fixed the test @patrickangle and myself talked about it today. We realized there's a missing callback for the _second_ call to `parent.didLayoutSubtree()`,  which is where the `resolve()` should actually go. 

Our changed implementation of `WI.View` will call it on the parent view for each layout of the subview:
2x subview layout  -> 2x `parent.didLayoutSubtree()`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js:
```diff
@@ -221,18 +240,28 @@ WI.ComputedStyleSection = class ComputedStyleSection extends WI.View
 
     applyFilter(filterText)
     {
+        if (this._filterText && !filterText)
+            this._filterCleared = true;
```
if you wanted to be a bit fancier you could
```
this._filterCleared ||= this._filterText && !filterText;
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
```diff
@@ -89,47 +89,43 @@ WI.ComputedStyleDetailsPanel = class ComputedStyleDetailsPanel extends WI.StyleD
-        case WI.ComputedStyleDetailsPanel.VariablesGroupingMode.Ungrouped:
-            styleGroups.set(this._variablesStyleSectionForGroupTypeMap.get(WI.DOMNodeStyles.VariablesGroupType.Ungrouped), this.nodeStyles.computedStyle);
-            break;
+            super.refresh(significantChange);
```
it seems like this is called in both paths.  should we maybe just move it to the top of this function and only call it once instead?

rcaliman-apple replied:
Fixed

dcrousso reviewed (approved):
nice work!  thanks for iterating :)

part of me wants to ask you to split out the `WI.View` changes into a separate change, but given how this is kinda the first _real_ use of it I suppose it's fine to keep together

rcaliman-apple committed:
Web Inspector: Elements Tab: Computed: variables section looks broken when the selected element has no CSS variables
https://bugs.webkit.org/show_bug.cgi?id=242704

Reviewed by Devin Rousso.

The original implementation made some wrong assumptions about the layout lifecycle
for CSS variable groups. This patch introduces a refactoring to correct that and
fix a bunch issues at once.

Context:

The properties section of the Computed panel is updated in response to the event
`WI.CSSStyleDeclaration.Event.PropertiesChanged` fired by
`WI.DOMNodeStyles.computedStyle` which is a `WI.CSSDeclaration` instance.
This allows it to dynamically reflect changes to any styles of the selected node.

There's an optimization in `WI.DOMNodeStyles.refresh() -> fetchedComputedStyle()`
where a change to styles of existing CSS rules isn't considered a `significantChange`.
This prevents a lot of needless re-layout in components which observe style refreshes.

The Properties section of the Computed panel listens for the event on `WI.DOMNodeStyles.computedStyle`
to address this optimization.

When grouping CSS variables, the groups were distinct `WI.CSSDeclaration` instances for clarity.
But they were cached and never updated in response to style changes on a node.
Because of this, the sections for variable groups in the Computed panel did not reflect
any changes: adding/removing variables, changing their type, etc.

The changes:

1) The logic for grouping variables by type is moved from `WI.DOMNodeStyles` to
`WI.CSSDeclaration`. This way, all sections in the Computed panel rely on a single
data source: `WI.DOMNodeStyles.computedStyle`.
It's the view's job to pick the subset of properties to render.

2) A new getter is added, `WI.CSSDeclaration.variablesForType`, to do this extra work
for all variable groups when consumers ask for them.

There can be a growing number of groups, so having separate iterations on every style
change over a potentially long list of properties is wasteful.
Plus, anything that doesn't match a type checker needs to go in the "Other" group.

3) `WI.ComputedStyleSection` is modified to obtain a list of properties to render
depending on the variables group type set on `WI.ComputedStyleSection.variablesGroupType`.
This subset of properties is then used in `WI.ComputedStyleSection.layout()`.

Introduced debouncing for layout in respose to `WI.DOMNodeStyles.Events.PropertiesChanged`
because it is expensive; throwing away all elements for properties and rebuilding from scratch.

4) Filtering is implemented. Sections that have no matches are hidden.

5) The test that checks CSS variables grouping is renamed and updated.

6) Modify `WI.View` to call `WI.View.didSubtreeLayout()` on the immediate parent
of the first dirty view that did layout on itself and all of its descendant views.
This allows a parent view to react in response to layout under its subtree.

This behavior is used to call `WI.ComputedStyleDetailsPanel.didSubtreeLayout()` where we can
toggle the visibility of the "Variables" section when the nested sections for variable groups are empty.

* LayoutTests/inspector/css/variablesForType-expected.txt: Renamed from LayoutTests/inspector/css/variableStylesByType-expected.txt.
* LayoutTests/inspector/css/variablesForType.html: Renamed from LayoutTests/inspector/css/variableStylesByType.html.
* LayoutTests/inspector/view/asynchronous-layout-expected.txt:
* LayoutTests/inspector/view/asynchronous-layout.html:
* LayoutTests/inspector/view/resources/test-view.js:
(TestPage.registerInitializer.WI.TestView):
(TestPage.registerInitializer.WI.TestView.prototype.get didLayoutSubtreeCount):
(TestPage.registerInitializer.WI.TestView.prototype.didLayoutSubtree):
(TestPage.registerInitializer):
* LayoutTests/inspector/view/synchronous-layout-expected.txt:
* LayoutTests/inspector/view/synchronous-layout.html:
* Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:
(WI.CSSStyleDeclaration):
(WI.CSSStyleDeclaration.prototype.variablesForType):
* Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
(WI.DOMNodeStyles):
(WI.DOMNodeStyles.prototype.refresh.fetchedComputedStyle):
(WI.DOMNodeStyles.prototype.get variableStylesByType): Deleted.
* Source/WebInspectorUI/UserInterface/Views/ComputedStyleDetailsPanel.js:
(WI.ComputedStyleDetailsPanel.displayNameForVariablesGroupType):
(WI.ComputedStyleDetailsPanel.prototype.refresh):
(WI.ComputedStyleDetailsPanel.prototype.layout):
(WI.ComputedStyleDetailsPanel.prototype.didLayoutSubtree):
(WI.ComputedStyleDetailsPanel.prototype._createVariablesStyleSection):
(WI.ComputedStyleDetailsPanel.prototype._removeVariablesStyleSection):
(WI.ComputedStyleDetailsPanel.prototype._renderVariablesStyleSectionGroup): Deleted.
* Source/WebInspectorUI/UserInterface/Views/ComputedStyleSection.js:
(WI.ComputedStyleSection.prototype.get variablesGroupType):
(WI.ComputedStyleSection.prototype.get isEmpty):
(WI.ComputedStyleSection.prototype.get properties):
(WI.ComputedStyleSection.prototype.layout):
(WI.ComputedStyleSection.prototype.detached):
(WI.ComputedStyleSection.prototype.applyFilter):
(WI.ComputedStyleSection.prototype._handlePropertiesChanged):
(WI.ComputedStyleSection):
* Source/WebInspectorUI/UserInterface/Views/View.js:
(WI.View.prototype.updateLayout):
(WI.View.prototype._layoutSubtree):
(WI.View._visitViewTreeForLayout):
(WI.View):

Canonical link: https://commits.webkit.org/253562@main
